unified_planning/engines/mixins/simulator.py

```diff
@@ -0,0 +1,131 @@
+        '''
+        Returns None if the event is not applicable in the given state, otherwise returns a new RWState,
+        which is a copy of the given state but the applicable effects of the event are applied; therefore
+        some fluent values are updated.
+        :param state: the state where the event formulas are calculated.
+        :param event: the event that has the information about the conditions to check and the effects to apply.
+        :return: None if the event is not applicable in the given state, a new RWState with some updated values
+         if the event is applicable.
+        '''
+        return self._apply(event, state)
+
+    def _apply(self, event: 'Event', state: 'up.model.RWState') -> Optional['up.model.RWState']:
+        raise NotImplementedError
+
+    def apply_unsafe(self, event: 'Event', state: 'up.model.RWState') -> 'up.model.RWState':
+        '''
+        Returns a new RWState, which is a copy of the given state but the applicable effects of the event are applied; therefore
+        some fluent values are updated.
+        IMPORTANT NOTE: Assumes that self.is_applicable(state, event) returns True
+        :param state: the state where the event formulas are evaluated.
+        :param event: the event that has the information about the effects to apply.
+        :return: A new RWState with some updated values.
+        '''
+        return self._apply_unsafe(event, state)
+
+    def _apply_unsafe(self, event: 'Event', state: 'up.model.RWState') -> 'up.model.RWState':
+        raise NotImplementedError
+
+    def get_applicable_events(self, state: 'up.model.ROState') -> Iterator['Event']:
+        '''
+        Returns a view over all the events that are applicable in the given State;
+        an Event is considered applicable in a given State, when all the Event condition
+        simplify as True when evaluated in the State.
+        :param state: the state where the formulas are evaluated.
+        :return: an Iterator of applicable Events.
+        '''
+        return self._get_applicable_events(state)
+
+    def _get_applicable_events(self, state: 'up.model.ROState') -> Iterator['Event']:
+        raise NotImplementedError
+
+    def get_events(self,
+                   action: 'up.model.Action',
+                   parameters: Union[Tuple['up.model.Expression', ...], List['up.model.Expression']]) -> List['Event']:
+        '''
+        Returns a list containing all the events derived from the given action, grounded with the given parameters.
+        :param action: the action containing the information to create the event.
+        :param parameters: the parameters needed to ground the action
+        :return: the List of Events derived from this action with these parameters.
+        '''
+        if action not in cast(up.model.Problem, self._problem).actions:
+            raise UPUsageError('The action given as parameter does not belong to the problem given to the SimulatorMixin.')
```

Comment:
This should be moved again in the sequential simulator since we cannot assume that self._problem is a `Problem`